### ISSUE TITLE
Problem with pagination when missing ``q`` GET parameter for ``SearchView``

### ISSUE THREAD

**Issue by jbzdak** [django-haystack/django-haystack#1466]
* [x] Tested with the latest Haystack release
* [x] Tested with the current Haystack master branch

## Actual behaviour

Suppose ``haystack.generic_views.SearchView`` is bound to ``/search`` url. 

When I enter the search url ``/search`` I get a paginated list of all indexed entities. When I enter ``/search?page=2`` I get error stating that no such page exists when in fact there is more than a single page available. 

After some digging I found that different code paths are executed in both cases. 

When I enter ``/search`` following things happen: 
*  ``SearchForm.is_valid`` returns ``False`` (form instance wasn't passed ``data`` parameter, and hence ``is_bound`` returns true, which makes the form invalid)
* ``SearchView.form_invalid`` is called and it returns full queryset
* Full queryset is paginated 

While when I enter ``/search?page=2``: 
* ``SearchForm.is_valid`` returns ``True`` 
* ``SearchView.form_valid`` is called
* which returns ``form.search()`` which in case of missing ``q`` parameter returns ``self.no_query_found()`` (which is empty by default)

## Expected behaviour 

I'd like the behavior to be consistent, either return (whole) queryset or return empty queryset in both cases. 

Probably it would be better to always trigger second behavior, as then it can be customized by overriding ``self.no_query_found()``. If that is the case, following [patch should fix this case](https://github.com/django-haystack/django-haystack/compare/master...jbzdak:jbzdak-patch?expand=1)

## Configuration

* Operating system version: Debian Linux (on docker that still uses Debian Linux Image)
* Search engine version: Most recent stable Whoosh 
* Python version: 3.5
* Django version: 1.10.4
* Haystack version: Both most recent stable and master

**Comment by adriancarayol** [comment]
A possible solution could be: http://127.0.0.1:8000/search/all/?page=1**&q=query**

**Comment by flavioamieiro** [comment]
I also ran into this and it took me a while to realize this was an inconsistent behavior. Thanks @jbzdak for documenting it here.

In case anyone stumbles here, as I did, I worked around this by creating my own `SearchForm` subclass and overriding the `no_query_found()` method as [described in the source code](https://github.com/django-haystack/django-haystack/blob/master/haystack/forms.py#L36-L45), which is where the thorough explanation in this issue pointed.

That being said, I still think this inconsistency should be fixed, by always calling `no_query_found()` when there is no 'q' parameter in the data.
